include/zephyr/device.h

```diff
@@ -950,6 +950,96 @@ BUILD_ASSERT(sizeof(device_handle_t) == 2, "fix the linker scripts");
+
+/**
+ * @def DEVICE_DT_API_GET
+ *
+ * @brief Get a <tt>const struct device*</tt> from a devicetree node
+ * identifier which implements the specified API type
+ *
+ * @details Returns a pointer to a device object created from a
+ * devicetree node, if any device which implements the specified API
+ * was allocated by a driver.
+ *
+ * If no such device was allocated, this will fail at linker time. If
+ * you get an error that looks like <tt>undefined reference to
+ * __device_dts_ord_<N></tt>, that is what happened. Check to make
+ * sure your device driver is being compiled, usually by enabling the
+ * Kconfig options it requires.
+ *
+ * @param node_id A devicetree node identifier
+ * @param api_type Type of the API
+ * @return A pointer to the device object created for that node
+ */
+#define DEVICE_DT_API_GET(node_id, api_type) \
+	(&DEVICE_DT_API_NAME_GET(node_id, api_type))
+
+/** @def DEVICE_DT_INST_API_GET
+ *
+ * @brief Get a <tt>const struct device*</tt> for an instance of a
+ * DT_DRV_COMPAT compatible
+ *
+ * @details This is equivalent to
+ * <tt>DEVICE_DT_API_GET(DT_DRV_INST(inst, api_type))</tt>.
+ *
+ * @param inst DT_DRV_COMPAT instance number
+ * @param api_type Type of the API
+ * @return A pointer to the device object created for that instance
+ */
+#define DEVICE_DT_INST_API_GET(inst, api_type) \
+	(DEVICE_DT_API_GET(DT_DRV_INST(inst), api_type))
+
+/**
+ * @def DEVICE_DT_API_DEFINE
+ *
+ * @brief Define an API for a device instance.
+ *
+ * @details This macro creates a device object from a previously defined device
+ * object for a specific API type. The new device object is placed in ROM in an
+ * iterable section, grouped by API type.
+ *
+ * To access the new device object to use with the API wrappers,
+ * @ref DEVICE_DT_API_GET is used.
+ *
+ * @note The device instance must be defined using @ref DEVICE_DT_DEFINE before
+ * using this macro
+ *
+ * @param node_id Node id of device instance which implements the API
+ * @param api_type Type of the API
+ * @param api_ptr Pointer to the API implementation
+ */
+#define DEVICE_DT_API_DEFINE(node_id, api_type, api_ptr)                        \
+	const Z_DECL_ALIGN(struct device)                                           \
+	DEVICE_DT_API_NAME_GET(node_id, api_type)                                   \
+	__attribute__((__section__(".z_device_api_" # api_type))) = {               \
+		.name = DEVICE_DT_GET(node_id)->name,                                   \
+		.config = DEVICE_DT_GET(node_id)->config,                               \
+		.api = (api_ptr),                                                       \
+		.state = DEVICE_DT_GET(node_id)->state,                                 \
+		.data = DEVICE_DT_GET(node_id)->data,                                   \
+		.handles = DEVICE_DT_GET(node_id)->handles,                             \
+		COND_CODE_1(CONFIG_PM_DEVICE, (.pm = DEVICE_DT_GET(node_id)->pm), ())   \
+	};
```

Comment:
The end goal is to move away from the DEVICE_DT_GET entirely, no generic device access. There will be a transition period where all legacy drivers and apps will work with generic devices, but all devices should be updated to use these new macros, DEVICE_DT_DEFINE will be updated to either require api_type, or have api_ptr removed. This way, all devices, multi-api or not will work the same way.

Reply:
> The end goal is to move away from the DEVICE_DT_GET entirely, no generic device access. There will be a transition period where all legacy drivers and apps will work with generic devices, but all devices should be updated to use these new macros, DEVICE_DT_DEFINE will be updated to either require api_type, or have api_ptr removed. This way, all devices, multi-api or not will work the same way.

So if the end goal is to modify all devices, I'd like to see the full picture before buying this approach. This is providing a simple solution, but eventually, it assumes we'll refactor everything because DEVICE_DT_GET is not enough. Will the final solution still require 2 struct devices?

Reply:
No, in the end, all device structs will be created by API type, DEVICE_DT_GET(node_id, ..., api_ptr, api_type) and additional struct devices will be created if DEVICE_DT_API_DEFINE is used. In ROM, there will be a section for initializing devices, and a section containing iterable sections with device structs sorted by API type.